building/trusted_applications.rst

```diff
@@ -299,3 +305,90 @@ sequence:
+   Never deploy an :ref:`optee_os` binary with this key in production. Instead replace this keys as soon as possible with a public key and keep the private part of the key offline, preferably on an HSM.
+
+.. note::
+
+   Currently only a single key for signing TAs is supported by :ref:`optee_os`.
+
+TAs are signed using the ``sign.py`` script referenced from ``ta/mk/ta_dev_kit.mk`` in :ref:`optee_os`. Its default behavior is to sign a compiled TA binary and attach the signature to form a complete TA for deployment. For **offline** signing, a three-step process is required: In a first step a digest of the compiled binary has to be generated, in the second step this digest is signed offline using the private key and finally in the third step the binary and its signature are stitched together into the full TA.
+
+Offline Signing of TAs
+++++++++++++++++++++++
+
+The TA dev kit does sign an application as last step of the linking process. For example, the file ``ta/arch/arm/link.mk`` in the :ref:`optee_os` source tree contains the statement
+
+.. code-block:: make
+
+    $(q)$(SIGN) --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0 \
+	    --in $$< --out $$@
+
+To avoid build errors when signing offline, this make script needs to be adopted. The detailed usage of ``sign.py`` is given below.
+
+Overall, offline signing is done with the following sequence of steps:
+
+0. (Preparation) Generate a 2048 bit RSA key for signing in a secure, offline environment. Extract the public key and copy it to the ``keys`` directory in the :ref:`optee_os` source tree. Adjust ``TA_SIGN_KEY`` for different file/path names. (Copy and) modify the ``link.mk`` file for the default linking step to produce a digest of the TA binary instead of the full TA.
+
+1. Manually (or with the modified linking script) generate a digest of the TA binary using
+
+.. code-block:: bash
+
+    sign.py generate-digest --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0
+
+2. Sign this digest offline, e.g. with OpenSSL
+
+.. code-block:: bash
+
+   base64 --decode digestfile | openssl pkeyutl -sign -inkey $TA_SIGN_KEY -pkeyopt digest:sha256 -pkeyopt rsa_padding_mode:pkcs1 | base64 > sigfile
+
+or with pkcs11-tool using a Nitrokey HSM
+
+.. code-block:: bash
+
+   echo "0000: 3031300D 06096086 48016503 04020105 000420" | xxd -c 19 -r > /tmp/sighdr
+   cat /tmp/sighdr $(base64 --decode digestfile) > /tmp/hashtosign
+   pkcs11-tool --id $key_id -s --login -m RSA-PKCS --input-file /tmp/hashtosign | base64 > sigfile
+
+3. Manually (or with an extra make target) stitch the TA together using
+
+.. code-block:: bash
+
+    sign.py stitch-ta --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0
+
+By default the UUID is taken as the base file name for all files. Different file names and paths can be set through additional options to ``sign.py``:
+
+.. code-block:: txt
```

Comment:
I recommend not duplicating the full help text for maintainability. It does not add much information here.